File: memoization/Construct/countConstruct.py
from typing import List


class Solution:
    def countConstruct(self, target:str, wordBank: List[str], memo = None) -> int:
        if memo == None:
            memo = {}

        if target in memo: return memo[target]
        if target == "": return 1

        constructCount = 0
        for word in wordBank:
            if target.startswith(word):
                suffix = target[len(word):]
                constructCount += self.countConstruct(suffix, wordBank, memo)

        memo[target] = constructCount            
        return constructCount
                
        
solution = Solution()
print(solution.countConstruct("", ["cat", "dog", "mouse"])) #expected: 1
print(solution.countConstruct("abcdef", ["ab", "abc", "cd", "def", "abcd"])) #expected: 1
print(solution.countConstruct("skateboard", ["sk", "ateboard", "skate", "board", "ska", "teboard", "boar"])) #expected: 3
print(solution.countConstruct("eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeef", ["e", "ee", "eee", "eeee", "eeeee"])) #expected: 0

File: Tabulation/Construct/countConstruct.js
const countConstruct = (target, wordBank) => {
    let table = Array(target.length + 1).fill(0);
    table[0] = 1;

    for (let i = 0; i <= target.length; i++) {
        const current = table[i];

        if (current) {
            for (let word of wordBank) {
                if (target.slice(i, i + word.length) === word) {
                    table[i + word.length] += 1;
                }
            }
        }
    }

    return table[target.length];
};


console.log(countConstruct("", ["cat", "dog", "mouse"])); // 1
console.log(countConstruct("abcdef", ["ab", "abc", "cd", "def", "abcd"])); // 1
console.log(countConstruct("skateboard", ["bo", "rd", "ate", "t", "ska", "sk", "boar"])); // 0
console.log(countConstruct("skateboard", ["sk", "ateboard", "skate", "board", "ska", "teboard", "boar"])); // 3
console.log(countConstruct("eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeef", ["e", "ee", "eee", "eeee", "eeeee"])); // 0